Update LVGL module to 9.2.0

faxe1008 opened the pull request:
Update the lvgl module to version 9.0.0. Commits still need to be cleaned, opening this for others to also test.

Things left to do:

- [x] Investigate the failing tests
- [x] Fix LV_Z_FLUSH_THREAD
- [x] Check samples on displays of every color format
- [x] [Color on RGB888 display](https://github.com/zephyrproject-rtos/zephyr/pull/68168#issuecomment-2038234021 )
- [x] Solution for monochrome pixel format
- [x] Add the OSAL layer`
- [x] Figure out what to do about DMA2D
- [x] Add the new demos

faxe1008 committed:
west.yml: Update lvgl module to v9.2.0

Update the west yaml to point to the new LVGL version.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
samples: modules: lvgl: Update CMakeLists

Update the CMakeLists for LVGL to reflect the changes done in v9.2.0.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
samples: modules: lvgl: Update Kconfig depedencies

With v9.1.0 the benchmark sample now has a Kconfig depedency on
LV_USE_DEMO_WIDGETS.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
tests: lib: gui: lvgl: Migrate to version 9.2

With the update to v9.2.0 several Kconfig symbols were renamed or removed.
Adjust the test accordingly. Also several constants were renamed, rename
those also.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
modules: lvgl: Update gluecode to v9.2

This patch updates the module gluecode to be compatible with LVGL version
9.2. This includes changes done to display and input driver initialization
  and draw buffer handling.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 commented:
Located the issue this was happening because of this Kconfig in upstream lvgl:
```
		config LV_TXT_BREAK_CHARS
			string "Can break (wrap) texts on these chars"
			default " ,.;:-_)]}"
```

faxe1008 committed:
samples: modules: lvgl: accelerometer_chart: Migrate to v9.2

This patch fixes the style size setter usage in the accelerometer_chart
demo since it now expects two dimensions. Also the newly named functions
which are typedefed by the api_map in LVGL are updated to anticipate their
deprecation.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

everedero commented:
Samples have weird colors on ST25DV. For instance Elena looks kind of blue:

![elena_blue_st25dv](https://github.com/zephyrproject-rtos/zephyr/assets/98658309/1519bea8-2470-4ca3-a4fc-3086769e8387)

faxe1008 commented:
@everedero  Thanks for testing it out :+1:, the `st25dv_mb1283_disco` seems to have the `RGB888` color format. Will check on this and get back to you.

faxe1008 committed:
modules: lvgl: Add zephyr OSAL implementation

This patch adds the OSAL implementation of dynamic thread creation, mutex
and thread syncronization primitive (semaphore) enabling the use of the
parallel rendering architecture provided with LVGL 9.0. To use it, set
`CONFIG_LV_Z_USE_OSAL` and your preferred dynamic thread stack allocation
method (pool or heap).

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
samples: modules: lvgl: demos: Migrate to v9.2

This patch changes the usage of LVGL functions which are typedefed in the
api map to their new version to anticipate their deprecation.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
samples: subsys: display: lvgl: Migrate to v9.2

This patch changes the functions unsed in the subsys demo that are part of
the api map to anticipate their deprecation.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
tests: lib: gui: lvgl: Add testcase for OSAL implementation

This patch adds a testcase adding the OSAL to ensure the implementation is
built during CI.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
samples: modules: lvgl: demos: Add new demos

Add the new demos of LVGL 9.2:
- flex_layout
- keypad_and_encoder
- render
- scroll
- multilang

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 commented:
@erwango I think I found the issue, can you please recheck with the current version?

faxe1008 commented:
Seems like software rendering for monochrome displays is not part of v9 just yet. There is a PR pending regarding this:
[https://github.com/lvgl/lvgl/pull/5800](https://github.com/lvgl/lvgl/pull/5800) but its grayscale. Ideally there would be support for `LV_COLOR_FORMAT_I1` which is the compact representation zephyr uses.

kartben commented:
Hi @faxe1008 -- what's the status of this for 3.7? We are of course past feature freeze but I *think* that if this can be ready for this week, it could still be worthwhile/possible to ask the TSC for an exception to get this in. Thanks!

faxe1008 commented:
Hey @kartben 
Thanks for offering a pathway to take this to 3.7 (and sorry for the noise, I accidentally marked it as ready for review). I am currently working on adding monochrome support to upstream which is mandatory imo, because I have encountered a lot of projects that use it. From my side this needs to be finished and be part of a v9.2. So from my side its highly unlikely that this would make it to the deadline.

faxe1008 commented:
Update support for monochrome SW color blending has been merged today 🥳 
[https://github.com/lvgl/lvgl/pull/6345#issuecomment-2238830684](https://github.com/lvgl/lvgl/pull/6345#issuecomment-2238830684)

diegoherranz commented:
Does it mean we're still in time for 3.7.0?

faxe1008 commented:
@diegoherranz no only blockers are bow merged, this will be post 3.7.

diegoherranz commented:
I guess so, but had to ask :)
Thanks.

faxe1008 commented:
Resolved merge conflicts, and tested this against latest upstream `daa7fefb3a4bd058f1c5b5166871085876d0ada4` (module repo PR needs to be adjusted or dropped if we start tracking master). There is still some weirdness going on in regards to include paths, needs to be sorted. With some patch in LVGL it builds, need to check if there is a bug in the upstream include logic or if something is up with the CMake.

faxe1008 committed:
modules: lvgl: Add support for NXP PXP engine

Adds support for the NXP PXP engine. LVGL changed
the Kconfig symbol from LV_USE_GPU_NXP_PXP to
LV_USE_PXP, but hal_nxp still relies on
LV_USE_GPU_NXP_PXP, so add a temporary symbol for it.
Additionally the drawing engines need to invalidate
ranges in the dcache, the method needs to be provided
via a lvgl_support.h file.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 commented:
Fixed the include issue, and added PXP support in the latest commit. Some weirdness in regards to the function naming that is needed by PXP impl. Raised [6588](https://github.com/lvgl/lvgl/issues/6588) for this. I suspect the changes got lost during a merge that were performed in here [5685](https://github.com/lvgl/lvgl/pull/5685/files). Anyways this can be done in a additional PR if this addition stretches the contents of the PR to far :^)

faxe1008 commented on modules/lvgl/include/lv_conf.h:
```diff
@@ -10,37 +10,33 @@
+#ifdef CONFIG_LV_Z_USE_OSAL
+#define LV_USE_OS            LV_OS_CUSTOM
+#define LV_OS_CUSTOM_INCLUDE "lvgl_zephyr_osal.h"
+#endif /* CONFIG_LV_Z_USE_OSAL */
```
Duplicate

faxe1008 commented:
Did some benchmarks in regards to RAM/ROM usage, since the resource usage has been a concern in the past:

| Sample                                  | 8.4.0 (RAM/ROM)    | 9.2.0 (RAM/ROM)    | 9.2.0 (Adjustments) (RAM/ROM) |
|-----------------------------------------|--------------------|--------------------|-------------------------------|
| samples/subsys/display/lvgl             | 688096 / 312509    | 688195 / 378316    | 683459 / 197123               |
| samples/modules/lvgl/accel              | 680270 / 264924    | 680361 / 318199    | 675585 / 155074               |
| samples/modules/lvgl/demo/music         | 1194797 / 987603   | 1195321 / 1042328  | 1190777 / 903005              |

The new version requires more RAM/ROM with the samples as they are currently. However with tweaking the samples just a bit I was able to bring them down (last column of the table):
```
CONFIG_LV_CONF_MINIMAL=y
CONFIG_LV_DRAW_SW_SUPPORT_ARGB8888=y
CONFIG_LV_DRAW_SW_SUPPORT_RGB565=n
CONFIG_LV_DRAW_SW_SUPPORT_RGB565A8=n
CONFIG_LV_DRAW_SW_SUPPORT_RGB888=n
CONFIG_LV_DRAW_SW_SUPPORT_XRGB8888=n
CONFIG_LV_DRAW_SW_SUPPORT_L8=n
CONFIG_LV_DRAW_SW_SUPPORT_AL88=n
CONFIG_LV_DRAW_SW_SUPPORT_A8=n
CONFIG_LV_DRAW_SW_SUPPORT_I1=n
```
But I would make those changes in a follow up PR, this one should be just about updating the module.

mmahadevan108 commented:
@0xFarahFl please review

faxe1008 commented:
I added the DNM label because I did some more in depth testing and found two issues:

- https://github.com/lvgl/lvgl/issues/6870, essentially with the v9.2.0 version some headers were made private and the PXP engine draw unit does not build anymore. This is a issue with upstream (or module code), so either we do not add PXP support for now (it was not supported with 8.4.0, so no regression).
- The monochrome rendering at the moment only works for htiled displays, were as the majority of displays intree are htiled in normal orientation. The `flush_cb` needs a conversion method for those with an intermediary buffer.

gschwaer commented on modules/lvgl/lvgl_display_16bit.c:
```diff
@@ -9,28 +9,24 @@
-	uint16_t *buf_xy = (uint16_t *)(buf + x * 2U + y * 2U * buf_w);
-	*buf_xy = lv_color_to16(color);
+#ifdef CONFIG_LV_COLOR_16_SWAP
+	lv_draw_sw_rgb565_swap(px_map, flush.desc.buf_size);
```
Hi, `lv_draw_sw_rgb565_swap()` takes the number of pixels as the second argument, not the number of bytes.
```suggestion
	lv_draw_sw_rgb565_swap(px_map, flush.desc.buf_size / sizeof(uint16_t));
```

faxe1008 replied:
Very good catch, thanks for bringing it up 👍

faxe1008 commented:
Updated the module PR:
- PXP now compiles, tests on hardware tbd.
- Removal of Kconfig workarounds

Missing:
- Conversion from htiled to vtiled for monochrome draw buffers.

faxe1008 committed:
samples: drivers: video: capture_to_lvgl: Migrate to LVGL v9.2

Adjust the Kconfig symbols of the capture to lvgl sample. LV_MEM_CUSTOM has
been removed in v9.0 and LV_USE_IMG has been changed to LV_USE_IMAGE.
Fix the changed fields for the image descriptor struct.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 committed:
boards: stm32h7b3i_dk: Rename LVGL DMA2D Kconfig symbols

Rename the defaulted DMA2D Kconfig symbols to match the new upstream
nomenclature.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

ngphibang commented on samples/drivers/video/capture_to_lvgl/src/main.c:
```diff
@@ -131,11 +131,10 @@ int main(void)
 	display_blanking_off(display_dev);
 
 	const lv_img_dsc_t video_img = {
-		.header.always_zero = 0,
```
Should put these changes in a separate commit as they are unrelated. This make the changes easier to be tracked of later.

faxe1008 replied:
They are in a seperate commit?
https://github.com/zephyrproject-rtos/zephyr/pull/68168/commits/68290e58bbffeb19f81fa4f1f586fb8ce3fda5c6
Or do you mean that the commit title should be adjusted?

faxe1008 replied:
Alright, I thought that since the goal of the overall commit was to fix the build issue I'd have both in one, given that the PR is already quite hefty in terms of commit count. But sure, can split that.

faxe1008 replied:
Changed the title and body, is this better?

ngphibang replied:
yes, thanks

erwango commented on boards/st/stm32h7b3i_dk/Kconfig.defconfig:
```diff
@@ -22,10 +22,10 @@ if LVGL
 	default y
 
-config LV_GPU_DMA2D_CMSIS_INCLUDE
+config LV_DRAW_DMA2D_HAL_INCLUDE
 	default "stm32h7xx.h"
```
Note: We'll need to move this to `soc/st/stm32/stm32h7x/Kconfig.defconfig` as it's common to whole STM32H7 series so it doesn't make sense to put in board config.

faxe1008 replied:
Make this part of this PR or a susequent one?

erwango replied:
No need to address that now as this is not related with this change. It can wait.

GeorgeCGV replied:
DMA2D support in LVGL 9.2 might need extra code to work.

GeorgeCGV replied:
If I understand [dma2d-support](https://docs.lvgl.io/master/details/integration/chip/stm32.html#dma2d-support) correctly to benefit from dma2d we need to use `LV_USE_DRAW_DMA2D_INTERRUPT ` with `LV_USE_OS` and call `lv_draw_dma2d_transfer_complete_interrupt_handler` when "DMA2D "transfer complete" global interrupt is received". Therefore, extra interrupt setup and handling code is required.

faxe1008 replied:
Enabling `LV_Z_USE_OSAL` is already possible and together with dma2d should already give some performance improvement. Its true that the interrupt setup is needed to avoid actively polling for completion.

Since I do not own a device with chromart I can not test this easily. If someone wants to help out on this I appreciate every bit :^)

Finomnis commented:
Might become relevant, once merged: #79936

Finomnis commented:
@faxe1008 does this have a chance to make it into `4.1`?

Finomnis commented:
A project of mine requires screen alpha (like: using LVGL to render an overlay over an FPGA provided Video Stream)

Therefore I added an `screen_transparency` LVGL Zephyr sample: https://github.com/faxe1008/zephyr/compare/update_lvgl_90...Finomnis:alpha_screen

It seems to fail, though. It seems to render the 320x240 native_sim screen in 24-line chunks, and there seem to be two issues:
- Screen does not get handled properly
- Data from the previous `write` call seems to get carried over

I'm not sure how much of that is the mistake of LVGL and how much is incorrect handling in the `display_sdl` driver, but I just wanted to leave that here:

![image](https://github.com/user-attachments/assets/7696449c-28d2-47a0-9cfe-cf964e247152)

kartben commented:
> A project of mine requires screen alpha (like: using LVGL to render an overlay over an FPGA provided Video Stream)
> 
> Therefore I added an `screen_transparency` LVGL Zephyr sample: https://github.com/faxe1008/zephyr/compare/update_lvgl_90...Finomnis:alpha_screen
> 
> It seems to fail, though. It seems to render the 320x240 native_sim screen in 24-line chunks, and there seem to be two issues:
> - Alpha does not get handled properly
> - Data from the previous `write` call seems to get carried over
> 
> I'm not sure how much of that is the mistake of LVGL and how much is incorrect handling in the `display_sdl` driver, but I just wanted to leave that here:
> 
> ![image](https://github.com/user-attachments/assets/7696449c-28d2-47a0-9cfe-cf964e247152)
> 

@Finomnis what are you trying to accomplish with this, this looks very suspicious? Looks like you've changed it for other samples as well but I fail to see why, and how this wouldn't cause a veryyyy long sleep?

https://github.com/faxe1008/zephyr/compare/update_lvgl_90...Finomnis:alpha_screen#diff-7fcee5f715ccfa3296cf8da91c973f88c95c8d54c91f2bbc4a25ab1d4c8ea9f2R64-R65

Finomnis commented:
@faxe1008 I added screen transparency support to the `display_sdl` driver.

Half the issues were fixed by that, but the weird duplication still exists.
I see that one directly in the `write` calls, so it does not stem from the `display_sdl` driver, but instead seems to be a zephyr-lvgl bug.

![image](https://github.com/user-attachments/assets/59c5913e-86c7-4b08-bba1-865e11487207)

Finomnis commented:
@faxe1008 Found the error. It's an LVGL bug.

```c
void lv_draw_buf_clear(lv_draw_buf_t * draw_buf, const lv_area_t * a)
{
    LV_ASSERT_NULL(draw_buf);

    const lv_image_header_t * header = &draw_buf->header;
    uint32_t stride = header->stride;

    if(a == NULL) {
        uint8_t * buf = lv_draw_buf_goto_xy(draw_buf, 0, 0);
        lv_memzero(buf, header->h * stride);
        return;
    }

    lv_area_t a_draw_buf;
    a_draw_buf.x1 = 0;
    a_draw_buf.y1 = 0;
    a_draw_buf.x2 = draw_buf->header.w - 1;
    a_draw_buf.y2 = draw_buf->header.h - 1;

    lv_area_t a_clipped;
    if(!lv_area_intersect(&a_clipped, a, &a_draw_buf)) return;
    if(lv_area_get_width(&a_clipped) <= 0) return;
    if(lv_area_get_height(&a_clipped) <= 0) return;

    uint8_t * buf = lv_draw_buf_goto_xy(draw_buf, a_clipped.x1, a_clipped.y1);
    uint8_t bpp = lv_color_format_get_bpp(header->cf);
    uint32_t line_length = (lv_area_get_width(&a_clipped) * bpp + 7) >> 3;
    int32_t y;
    for(y = a_clipped.y1; y <= a_clipped.y2; y++) {
        lv_memzero(buf, line_length);
        buf += stride;
    }
}
```

`lv_area_intersect` fails. The problem is, this function assumes that the x/y coordinates of the draw buffer are `0,0`, which is incorrect for partial buffers.

Finomnis commented:
@faxe1008 Created a bug report. (https://github.com/lvgl/lvgl/issues/7268)

Unsure if that affects you in any way though, as none of this seems to be related to the Zephyr LVGL integration.

faxe1008 commented:
> does this have a chance to make it into `zephyr 4.1`?

I really hope so, yes.

> Found the error. It's an LVGL bug.
> 
> Created a bug report. ([lvgl/lvgl#7268](https://github.com/lvgl/lvgl/issues/7268))
> 
> Unsure if that affects you in any way though, as none of this seems to be related to the Zephyr LVGL integration.

Thanks for investigating and raising the upstream issue (and even providing a fix for it :heart: ). With the fix merged we only need to point the module to new new rev :^)

faxe1008 commented:
> do we need to bump this somewhere to get it reviewed? Cause there isn't much activity on the lvgl repo so far.

Yeah the turn-around time for PRs on upstream is pretty high. This is part of the reason why I have been chipping away at this since start of this year. I am not aware of any "escalation" paths or something similar. There is a support forum but do not know if this is the place for it. I have not used it personally though.

faxe1008 commented on modules/lvgl/lvgl_display.c:
```diff
@@ -65,50 +66,46 @@ void lvgl_rounder_cb(lv_disp_drv_t *disp_drv, lv_area_t *area)
+					display);
 		break;
 	case PIXEL_FORMAT_MONO01:
 	case PIXEL_FORMAT_MONO10:
```
This causes issues in the `tests/lib/gui/lvgl` for the mono test case `libraries.gui.lvgl.mono`

`west build -p -b native_sim/native tests/lib/gui/lvgl -T libraries.gui.lvgl.mono`

Reason is that there is a discrepency between the usage of the LV_COLOR_DEPTH kconfigs of LVGL and the pixel ofrmat that one assigns to the display. In the testcase we set `CONFIG_LV_COLOR_DEPTH_1` but this does not change the actual pixel format of the display. Therefore with the new optimization for monochrome displays (1 bit per pixel instead of a whole byte) this fails internal buffer length checks of LVGL.  Raised this PR https://github.com/zephyrproject-rtos/zephyr/pull/81263 so we can add overlays for the tests to ensure that the code paths are run as expected.

faxe1008 replied:
Essentially this is what the test code now does:
https://github.com/zephyrproject-rtos/zephyr/blob/f9981e34b4b4d3254e774986082b097fbf57aafc/tests/lib/gui/lvgl/src/main.c#L118

Before calling the init manually the display pixel format is set. Static allocation falls flat here, because your buffer needs to accomodate for a possible runtime switch. This means if you set `LV_Z_BITS_PER_PIXEL` high and switch the pixel format to a smaller one you waste some memory on your static draw buffers. But there is little you can do about.

danieldegrasse replied:
Sure, that makes sense about static allocation. I think what I'm confused about is if is possible to generically set the display pixel format before `lvgl_init` runs, or if we need to use this workaround to explicitly delay init.

IE could we modify our LVGL porting code so that the display pixel format is set before we call `lvgl_init`, and avoid the assertion?

faxe1008 replied:
Not sure if I understand the suggested change.

> IE could we modify our LVGL porting code so that the display pixel format is set before we call lvgl_init, and avoid the assertion?

Just to make sure that we are on the same page 😄 :
The display driver is initialized with the `POST_KERNEL` init priority. Most in-tree drivers have a default current `pixel_format` set after init is run. For some device tree bindings the initial `pixel_format` is configurable via device tree.
After this the LVGL init code is run inside the `SYS_INIT` with `APPLICATION` init priority.

So there are two cases:

- The display has the "correct" pixel format set after init. In this case static allocation and everything else will work just fine. The `pixel_format` is "visible" to the LVGL init code during `SYS_INIT`
- The display has a "wrong" pixel format configured. In this case one would need to set `LV_Z_AUTO_INIT=n` and somewhere within application code do the following:
```
display_set_pixel_format(dev, PIXEL_FORMAT_MONO01);
lvgl_init();
```
This will work with the dynamic allocation just fine, with the static one there is the potential that you have to over-allocate.

Does your question mean that you want the LVGL init code to set the displays `pixel_format`? So having a Kconfig property like `LV_Z_DISPLAY_PIXEL_FORMAT` or the like and set the pixel format based on that during the init?

danieldegrasse replied:
> Does your question mean that you want the LVGL init code to set the displays pixel_format? So having a Kconfig property like LV_Z_DISPLAY_PIXEL_FORMAT or the like and set the pixel format based on that during the init?

Honestly, I think this *might* make sense. LVGL already has a good idea of the pixel format it expects from the display based on the choice of `CONFIG_LV_COLOR_DEPTH`, right?

- `LV_COLOR_DEPTH_32`: `PIXEL_FORMAT_ARGB_8888`
- `LV_COLOR_DEPTH_16`: `PIXEL_FORMAT_RGB_565` if `LV_COLOR_16_SWAP == y`, otherwise `PIXEL_FORMAT_BGR_565`
- `LV_COLOR_DEPTH_1`: check if pixel format is `PIXEL_FORMAT_MONO01` or `PIXEL_FORMAT_MONO_10`, if not set the format to `MONO01`.

The `MONO10` versus `MONO01` does leave some ambiguity, which might be an issue. Otherwise though I think this would be useful- LVGL would be able to detect during init if a display could not support the chosen pixel format

faxe1008 replied:
Hmm might make sense for an enhancement, in the future. With the current changeset we are not painting ourselves into a corner. The delayed init was request by some people using deferred-init with their displays so it's needed either way. I want to get this one merged first and then think about enhancements. LVGL also supports multiple screens so a refactoring of sorts will be needed for that.

Finomnis commented:
@faxe1008 https://github.com/lvgl/lvgl/pull/7269 got merged

faxe1008 commented:
@Finomnis I forwarded it to the module PR, thank you very much in fixing this :^)

Finomnis commented:
#81184 got merged, I think that one collides with this PR

faxe1008 committed:
samples: modules: lvgl: transparency: Migrate to v9.2

Removes the LV_COLOR_SCREEN_TRANSP Kconfig option as its not part of the
modules Kconfig anymore and renames the functions according to the new v9.2
API.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

Finomnis commented:
Updated `samples/modules/lvgl/screen_transparency/src/main/c`:

```diff
diff --git a/samples/modules/lvgl/screen_transparency/src/main.c b/samples/modules/lvgl/screen_transparency/src/main.c
index 64494ffdaed..cf859fac2e9 100644
--- a/samples/modules/lvgl/screen_transparency/src/main.c
+++ b/samples/modules/lvgl/screen_transparency/src/main.c
@@ -21,14 +21,14 @@ static void initialize_gui(void)
        lv_obj_t *label;
 
        /* Configure screen and background for transparency */
-       lv_obj_set_style_bg_opa(lv_scr_act(), LV_OPA_TRANSP, LV_PART_MAIN);
-       lv_disp_set_bg_opa(NULL, LV_OPA_TRANSP);
-       lv_obj_set_style_bg_color(lv_scr_act(), lv_color_hex(0x000000), LV_PART_MAIN);
+       lv_obj_set_style_bg_opa(lv_screen_active(), LV_OPA_TRANSP, LV_PART_MAIN);
+       lv_obj_set_style_bg_opa(lv_layer_bottom(), LV_OPA_TRANSP, LV_PART_MAIN);
+       lv_obj_set_style_bg_color(lv_screen_active(), lv_color_hex(0x000000), LV_PART_MAIN);
 
        /* Create a label, set its text and align it to the center */
-       label = lv_label_create(lv_scr_act());
+       label = lv_label_create(lv_screen_active());
        lv_label_set_text(label, "Hello, world!");
-       lv_obj_set_style_text_color(lv_scr_act(), lv_color_hex(0xff00ff), LV_PART_MAIN);
+       lv_obj_set_style_text_color(lv_screen_active(), lv_color_hex(0xff00ff), LV_PART_MAIN);
        lv_obj_align(label, LV_ALIGN_CENTER, 0, 0);
 }
 
@@ -60,11 +60,11 @@ int main(void)
 
        initialize_gui();
 
-       lv_task_handler();
+       lv_timer_handler();
        display_blanking_off(display_dev);
 
        while (1) {
-               uint32_t sleep_ms = lv_task_handler();
+               uint32_t sleep_ms = lv_timer_handler();
 
                k_msleep(MIN(sleep_ms, INT32_MAX));
        }
```

faxe1008 commented:
@Finomnis updated the PR, thanks for catching it :^)

Finomnis commented:
Except of #81263, is this change feature complete? No other open known issues?

faxe1008 commented:
@Finomnis yes, everything else should be done. Some more tests for monochrome displays with real hardware would be good though.

Finomnis commented:
@jpeeters I think this PR is at the `master` branch, already newer than the `9.2.2` release.

jpeeters commented:
@Finomnis Sorry for asking but where can I find the integrated commit sha1 from LVGL? I am looking for if STM32 DMA2D and NemaGFX is ready to use in the update on zephyr?

faxe1008 commented:
@jpeeters you can find the commits in the module PR. This is the current HEAD.
https://github.com/faxe1008/zephyr-lvgl/commit/1eb2d3aee876b270b483caa7e9e65acce2d17af1

pdgendt commented on modules/lvgl/lvgl.c:
```diff
@@ -58,39 +66,33 @@ static uint8_t buf1[BUFFER_SIZE]
-	case 'E':
+	switch (level) {
+	case LV_LOG_LEVEL_ERROR:
 		LOG_ERR("%s", buf + strlen("[Error] "));
```
Since you're touching this, use compile time data instead of runtime data.

```suggestion
		LOG_ERR("%s", buf + (sizeof("[Error] ") - 1));
```

faxe1008 replied:
Good idea 👍

faxe1008 replied:
Done.

pdgendt commented:
CC @uLipe as you've joined the LVGL collaborators.

faxe1008 commented:
Encountered bug in DMA2D code, upstream PR https://github.com/lvgl/lvgl/pull/7456

uLipe commented:
@faxe1008 @pdgendt just approved on upstream, just need one more approval to merge it.

faxe1008 commented:
After quite some bit fiddeling I found the last remaining issue in mono rendering:
![image](https://github.com/user-attachments/assets/e64a6731-5221-4bdf-ae24-f34025130db2)

Since I have run out of things to fix, this one is now ready for thorough review (and hopefuly merge) @pdgendt @kartben @diegoherranz @danieldegrasse @uLipe

pdgendt commented on modules/lvgl/lvgl_zephyr_osal.c:
```diff
@@ -0,0 +1,150 @@
+	return LV_RESULT_OK;
+}
+
+lv_result_t lv_thread_sync_delete(lv_thread_sync_t *sync)
+{
+	ARG_UNUSED(sync);
```
Not sure how this is used, but maybe the semaphore could be reset here?

faxe1008 replied:
From the other OSALS what I saw was that this was meant to be used if the underlying type allocates further ressources in the `init` function. Resetting could be done, but if you were to re-init the semaphore the count is reset anyways, right?

pdgendt commented on modules/lvgl/Kconfig:
```diff
@@ -128,15 +130,45 @@ config LV_Z_AREA_Y_ALIGNMENT_WIDTH
+	  Use the Zephyr LVGL OSAL to enable parallel rendering
+	  pipelines.
+
+config LV_USE_PXP
+	bool
```
Should this be promptless? Doesn't it make sense to allow disabling the PXP?

faxe1008 replied:
Yeah the reason for this symbol is that LVGL v9 https://github.com/lvgl/lvgl/blob/master/Kconfig#L384 names it `LV_USE_PXP`, former versions used `LV_USE_GPU_NXP_PXP` and at some point made its way into hal_nxp https://github.com/zephyrproject-rtos/hal_nxp/blob/6ac83c6ba14830871e19db7bd4c30af8b577fce2/mcux/hal_nxp.cmake#L218.

So for backwards compatibility to the hal layer this was added as a forward declared prompt so that the old symbol can be enabled:
```
config LV_USE_GPU_NXP_PXP
	bool
	default y if LV_USE_PXP
```

pdgendt reviewed (dismissed):
Nice work!

uLipe commented on modules/lvgl/lvgl.c:
```diff
@@ -222,28 +246,26 @@ static int lvgl_init(void)
-	if (err != 0) {
-		return err;
+	display = lv_display_create(disp_data.cap.x_resolution, disp_data.cap.y_resolution);
+	if (!display) {
```
I think failure on the display creation could be reported as -ENOMEM or -ENODEV since -ENOTSUP tends to indicate the actual function is not supported, which seems not the case.

faxe1008 replied:
The error code is better true 👍 , fixed.

uLipe commented on modules/lvgl/lvgl.c:
```diff
@@ -253,4 +275,6 @@ static int lvgl_init(void)
 	return 0;
 }
 
+#ifdef CONFIG_LV_Z_AUTO_INIT
```
I'm wondering if would be good also to make the priority of the auto init something that can be configured.

faxe1008 replied:
Sure why not does not hurt to have this in case you have some intricate setup where you need to manage init priorities within the application level. Added this.

uLipe commented on modules/lvgl/lvgl_zephyr_osal.c:
```diff
@@ -0,0 +1,150 @@
+{
+	lv_thread_entry entry_cb = (lv_thread_entry)cb;
+
+	entry_cb(user_data);
```
Even though I think entry_cb will not be NULL at any case, if this occurs, it would be hard to debug, said that , it would not useful here to something like this:

```
if(likely(entry_cb != NULL)) {
      entry_cb(user_data);
} else {
     //put an error type log here. 

}
```

faxe1008 replied:
I added an __ASSERT for this instead because imo its a bug if you want to start a thread without an entry function.

uLipe reviewed (commented):
Initial comments. but overall looks good!

pdgendt commented on modules/lvgl/Kconfig:
```diff
@@ -128,15 +130,51 @@ config LV_Z_AREA_Y_ALIGNMENT_WIDTH
+	int "Default init priority for LVGL"
+	default 90
+	depends on LV_Z_AUTO_INIT
+	  Priority used for the automatic initialization of LVGL.
```
Missing `help`

```suggestion
	help
	  Priority used for the automatic initialization of LVGL.
```

faxe1008 replied:
ups, yeah fixed :^)

kartben commented:
@faxe1008 awesome work, thank you so much!

Finomnis commented:
It is done!! Crazy. Thanks, @faxe1008

Finomnis commented:
@faxe1008 Missing entry in changelog, from what I can see

faxe1008 commented:
@Finomnis I do this typically during feature freeze of the next release :^)
